### PR TITLE
Fix query failure race in BroadcastOutputBuffer

### PR DESCRIPTION
**Pull request by dain**
Client buffer added to a broadcast output buffer were being allowed to be completed
cleanly. This means, if a broadcast buffer is failed before any client buffers are
added, the client stage will see a finished buffer and will initiate stage shutdown.
This results in a race between the propigation of the failure, and the client stage
shutting down.  In rare cases this can cause a query to incorrectly report a susccess
state.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/buffer/BroadcastOutputBuffer.java [hunk 1]
```diff
@@ -299,23 +299,28 @@ private synchronized ClientBuffer getBuffer(OutputBufferId id)
...
+            // add initial pages
+            buffer.enqueuePages(initialPagesForNewBuffers);
 
-        // update state
-        if (!state.get().canAddPages()) {
-            buffer.setNoMorePages();
-        }
+            // update state
+            if (!state.canAddPages()) {
+                // DO NOT set no more pages if FAILED
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/execution/buffer/BroadcastOutputBuffer.java]: This comment is a little confusing, since state is guaranteed to be NOT FAILED at this point. I'd rephrase as "set no more pages only if not FAILED" and maybe add a little more context.

**Comment by electrum** [comment]
Typo "propigation" in commit description
